src/nwb_benchmarks/core/_streaming.py

```diff
@@ -0,0 +1,96 @@
+) -> Tuple[h5py.File, HTTPFile]:
+    """Load the raw HDF5 file from an S3 URL using fsspec without a cache; does not formally read the NWB file."""
+    reset_lock()
+    fsspec.get_filesystem_class("https").clear_instance_cache()
+    filesystem = fsspec.filesystem("https")
+
+    byte_stream = filesystem.open(path=s3_url, mode="rb")
+    file = h5py.File(name=byte_stream)
+    return (file, byte_stream)
+
+
+def read_hdf5_nwbfile_fsspec_no_cache(
+    s3_url: str,
+) -> Tuple[pynwb.NWBFile, pynwb.NWBHDF5IO, h5py.File, HTTPFile]:
+    """Read an HDF5 NWB file from an S3 URL using fsspec without a cache."""
+    (file, byte_stream) = read_hdf5_fsspec_no_cache(s3_url=s3_url)
+    io = pynwb.NWBHDF5IO(file=file, load_namespaces=True)
+    nwbfile = io.read()
+    return (nwbfile, io, file, byte_stream)
+
+
+def read_hdf5_remfile(s3_url: str) -> Tuple[h5py.File, remfile.File]:
+    """Load the raw HDF5 file from an S3 URL using remfile; does not formally read the NWB file."""
+    byte_stream = remfile.File(url=s3_url)
+    file = h5py.File(name=byte_stream)
+    return (file, byte_stream)
+
+
+def read_hdf5_nwbfile_remfile(s3_url: str) -> Tuple[pynwb.NWBFile, pynwb.NWBHDF5IO, h5py.File, remfile.File]:
+    """Read an HDF5 NWB file from an S3 URL using the ROS3 driver from h5py."""
+    (file, byte_stream) = read_hdf5_remfile(s3_url=s3_url)
+    io = pynwb.NWBHDF5IO(file=file, load_namespaces=True)
+    nwbfile = io.read()
+    return (nwbfile, io, file, byte_stream)
+
+
+def robust_ros3_read(
+    command: Callable,
+    max_retries: int = 20,
+    command_args: Union[list, None] = None,
+    command_kwargs: Union[dict, None] = None,
+) -> Any:
+    """
+    Attempt the command (usually acting on an S3 IO) up to the number of max_retries using exponential backoff.
+
+    Usually a good idea to use this to wrap any operations performed using the ROS3 driver.
+    """
+    command_args = command_args or []
+    command_kwargs = command_kwargs or dict()
+    for retry in range(max_retries):
+        try:
+            result = command(*command_args, **command_kwargs)
+            return result
+        except Exception as exception:
+            #  'cannot curl request' can show up in potentially many different error types
+            if "curl" not in str(exception):
+                raise exception
+            time.sleep(0.1 * 2**retry)
+
+    raise TimeoutError(f"Unable to complete the command ({command.__name__}) after {max_retries} attempts!")
+
+
+def read_hdf5_ros3(s3_url: str) -> h5py.File:
+    """Load the raw HDF5 file from an S3 URL using ROS3 driver; does not formally read the NWB file."""
+    ros3_form = s3_url.replace("https://dandiarchive.s3.amazonaws.com", "s3://dandiarchive")
+    aws_region = bytes("us-east-2", "ascii")  # TODO: generalize this as an argument if necessary
+    file = robust_ros3_read(
+        command=h5py.File, command_kwargs=dict(name=ros3_form, driver="ros3", aws_region=aws_region)
+    )
+    return file
```

Comment:
I'm wondering whether this should  be called `open_hdf5_ros3` instead? As far as I can tell, the function seems to open an HDF5 file by creating an `h5py.File` instance but it doesn't seem to actually read.

Reply:
It at least reads the identifier of the file; I'm not sure if there's any other very top-level metadata it grabs at the same time